### PR TITLE
Client add option set

### PR DESCRIPTION
**Pull request by outscale-nts**

### CODE REVIEW

File: api/client/sg.cc [hunk 1]
```diff
@@ -422,20 +422,12 @@ static void sub_sg_rule_del_help(void) {
...
-        sub_sg_rule_add_help();
-        return 0;
-    }
-
-    if (argc <= 5) {
-        sub_sg_rule_del_help();
-        return 1;
-    }
+static bool check_hash(const string &param) {
+    return param.find("-") == 0;
```
Comment (outscale-mgo) [inline api/client/sg.cc]: s/check_hash/is_hash/
you use "c" so a char *, but find can take a char in parameter, so you could simply use 'c', a char
`return param.find('-') ==  string::npos`

http://www.cplusplus.com/reference/string/string/find/

File: api/client/sg.cc [hunk 2]
```diff
@@ -422,20 +422,12 @@ static void sub_sg_rule_del_help(void) {
...
+}
 
-    string sg = string(argv[4]);
-    string hash = string(argv[5]);
+static int sub_sg_hashed_rule_del(string sg, string hash,
```
Comment (outscale-mgo) [inline api/client/sg.cc]:
```
static int sub_sg_hashed_rule_del(const string &sg, const string &hash, const GlobalOptions &options)
```
https://google.github.io/styleguide/cppguide.html#Use_of_const

File: api/client/sg.cc [hunk 3]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+            "  }"
+            "}";
+
+    proto::Messages res;
+    return request(req, &res, options, false) || check_request_result(res);
+}
+static int sub_sg_rule_del(int argc, char **argv,
+                            const GlobalOptions &options) {
+    if (argc >= 5 && strcmp(argv[4], "help") == 0) {
+        sub_sg_rule_add_help();
```
Comment (outscale-mgo) [inline api/client/sg.cc]: s/sub_sg_rule_add_help/sub_sg_rule_del_help/

File: api/client/sg.cc [hunk 4]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+            "    }"
+            "  }"
+            "}";
+
+    proto::Messages res;
+    return request(req, &res, options, false) || check_request_result(res);
+}
+static int sub_sg_rule_del(int argc, char **argv,
+                            const GlobalOptions &options) {
+    if (argc >= 5 && strcmp(argv[4], "help") == 0) {
```
Comment (outscale-mgo) [inline api/client/sg.cc]:
```
if ((argc >= 5 && strcmp(argv[4], "help") == 0) || argc <= 5)
```

File: api/client/sg.cc [hunk 5]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        sub_sg_rule_add_help();
+        return 0;
+    }
+
+    if (argc <= 5) {
```
Comment (outscale-mgo) [inline api/client/sg.cc]: and this block is now useless

File: api/client/sg.cc [hunk 6]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        sub_sg_rule_del_help();
+        return 1;
+    }
+    string sg = argv[4];
+    string param = argv[5];
```
Comment (outscale-mgo) [inline api/client/sg.cc]: string sg(argv[4]);
string param(argv[5]);

File: api/client/sg.cc [hunk 7]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    if (check_hash(param) == false) {
+        return sub_sg_hashed_rule_del(sg, param, options);
+    }
+    return sub_sg_param_rule_del(argc, argv, options);
```
Comment (outscale-jju) [inline api/client/sg.cc]: May `sub_sg_hashed_rule_del` and `sub_sg_param_rule_del` have the same signature ?

File: api/client/sg.cc [hunk 8]
```diff
@@ -422,20 +422,12 @@ static void sub_sg_rule_del_help(void) {
     global_parameter_help();
```
Comment (outscale-jju) [inline api/client/sg.cc]: need documentation update

File: api/client/sg.cc [hunk 9]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+        "  message_0 {"
+        "    request {"
+        "      sg_rule_del {"
+        "        sg_id: \"" + opts.sg + "\""
+        "        rule {"
+        "          direction: " + opts.direction +
+        "          protocol: " + to_string(opts.proto);
+    if (opts.has_port_start) {
+        req += "   port_start: " + to_string(opts.port_start) +
+            "   port_end: " + to_string(opts.port_end);
```
Comment (outscale-jju) [inline api/client/sg.cc]: indentation fail

File: api/client/sg.cc [hunk 10]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    }
+    if (opts.cidr.length())
+        req += "   cidr { " + opts.cidr + " }";
+    else if (opts.sg_members.length())
+            req += "   security_group: \"" + opts.sg_members + "\"";
```
Comment (outscale-jju) [inline api/client/sg.cc]: indentation fail

File: api/client/sg.cc [hunk 11]
```diff
@@ -481,7 +473,61 @@ static int sub_sg_rule_del(int argc, char **argv,
...
+    req +=
+            "        }"
+            "      }"
+            "    }"
+            "  }"
+            "}";
```
Comment (outscale-jju) [inline api/client/sg.cc]: indentation fail, does not match opening braces indentation

File: api/client/sg.cc (hunk 7 above)
Comment (outscale-jju) [inline api/client/sg.cc]: `check_hash(param) == false` => `!check_hash(param)`

Comment (outscale-jju) [inline api/client/sg.cc]: single statement in if => remove braces

File: api/client/nic.cc [hunk 12]
```diff
@@ -271,6 +271,37 @@ static int sub_nic_sg_add(int argc, char **argv,
     return sg_update(nic, all_sg, options);
 }
 
+static void sub_nic_sg_set_help(void) {
+    cout << "usage: butterfly nic sg set NIC SG [SG...] [options...]" << endl;
+    cout << endl << "Set security groups from a vnic" << endl;
```
Comment (outscale-jju) [inline api/client/nic.cc]: Would say: "Overwrite security group listing of a vnic"

Comment (outscale-jju) [inline api/client/nic.cc]: Also, would rewrite commit title to `client: add new sub-command 'nic sg set'` because `set` is ambiguous in your commit title

**Comment by outscale-jju** [comment]
rebased and merged

**Pull request closed without merging**